src/main/java/rip/bolt/lobby/manager/NPCManager.java

```diff
@@ -0,0 +1,188 @@
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.player.PlayerInteractEntityEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import rip.bolt.lobby.BoltLobby;
+import rip.bolt.lobby.api.definitions.BoltMatch;
+import rip.bolt.lobby.config.Config;
+import rip.bolt.lobby.events.MatchUpdateEvent;
+import rip.bolt.lobby.manager.servers.CircleLocationSpreader;
+import rip.bolt.lobby.manager.servers.NPCLocator;
+import rip.bolt.lobby.manager.servers.PlayerCountManager;
+import rip.bolt.lobby.manager.servers.ServerSlot;
+import rip.bolt.lobby.manager.servers.ServerStatus;
+import rip.bolt.lobby.utils.NMSHacks;
+import rip.bolt.lobby.utils.OnlinePlayerMapAdapter;
+import rip.bolt.lobby.utils.ViaUtils;
+
+public class NPCManager implements Listener {
+
+  private final BoltLobby lobby;
+
+  private Config.Npc npcConfig;
+  private NPCLocator locator;
+  private final PlayerCountManager playerCountUpdater;
+
+  private final long cooldown = Duration.ofSeconds(2).toMillis();
+  private final OnlinePlayerMapAdapter<Long> cooldowns;
+
+  public NPCManager(BoltLobby boltLobby, Config.Npc npcConfig) {
+    this.lobby = boltLobby;
+    this.npcConfig = npcConfig;
+
+    this.cooldowns = new OnlinePlayerMapAdapter<>(lobby);
+
+    this.locator = createLocator();
+
+    this.playerCountUpdater = new PlayerCountManager(this.lobby, this);
+  }
+
+  @EventHandler
+  public void onMatchUpdate(MatchUpdateEvent event) {
+    BoltMatch match = event.getMatch();
+    ServerStatus status = locator.getSlot(match).flatMap(ServerSlot::getServerStatus).orElse(null);
+    if (status == null) createStatue(match);
+    else status.updateMatch(match);
+  }
+
+  @EventHandler(priority = EventPriority.LOWEST)
+  public void onEntityDamage(EntityDamageByEntityEvent event) {
+    if (!(event.getDamager() instanceof Player)) return;
+    Player player = (Player) event.getActor();
+
+    event.setCancelled(validateClick(player, event.getEntity()));
+  }
+
+  @EventHandler(priority = EventPriority.LOWEST)
+  public void onPlayerInteractEntity(PlayerInteractEntityEvent event) {
+    validateClick(event.getPlayer(), event.getRightClicked());
+  }
+
+  @EventHandler(priority = EventPriority.LOWEST)
+  public void onEntityDamage(EntityDamageEvent event) {
+    if (isStatusEntity(event.getEntity())) event.setDamage(0);
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onPlayerJoin(PlayerJoinEvent event) {
+    Player player = event.getPlayer();
+
+    Bukkit.getScheduler()
+        .scheduleSyncDelayedTask(
+            this.lobby,
+            () -> {
+              if (!(ViaUtils.enabled() && ViaUtils.isLegacy(player)) && ViaUtils.isReady(player))
+                return;
```

Comment:
```suggestion
              if (!player.isOnline() || !ViaUtils.isLegacy(player)) return;
```

Just simplify it, make isLegacy return false if viautils isn't enabled